File: Algorithms/HeapSort.java
import java.util.Arrays;

class HeapSort {

    // Main function to sort an array using heap sort
    public static void heapSort(int[] array) {
        int n = array.length;

        // Build a max heap
        for (int i = n / 2 - 1; i >= 0; i--) {
            heapify(array, n, i);
        }

        // One by one extract elements from heap
        for (int i = n - 1; i >= 0; i--) {
            // Move the current root (maximum) to the end
            int temp = array[0];
            array[0] = array[i];
            array[i] = temp;

            // Call max heapify on the reduced heap
            heapify(array, i, 0);
        }
    }

    // To heapify a subtree rooted with node i, which is an index in array[].
    // n is the size of the heap
    private static void heapify(int[] array, int n, int i) {
        int largest = i; 
        int left = 2 * i + 1; 
        int right = 2 * i + 2; 
        // If left child is larger than root
        if (left < n && array[left] > array[largest]) {
            largest = left;
        }

        // If right child is larger than largest so far
        if (right < n && array[right] > array[largest]) {
            largest = right;
        }

        // If largest is not root
        if (largest != i) {
            int swap = array[i];
            array[i] = array[largest];
            array[largest] = swap;

            // Recursively heapify the affected subtree
            heapify(array, n, largest);
        }
    }

    // Main method to test the heap sort implementation
    public static void main(String[] args) {
        int[] arr = {12, 11, 13, 5, 6, 7};
        System.out.println("Original array: " + Arrays.toString(arr));

        heapSort(arr);

        System.out.println("Sorted array: " + Arrays.toString(arr));
    }
}

File: Algorithms/Heap_sort.cpp
#include <iostream>
#define MAX 100
using namespace std;

void heap_sort(int arr[], int size);
void buildHeap(int arr[], int size);
int del_root(int arr[], int *size);
void restoreDown(int arr[], int i, int size);
void display(int arr[], int n);

int main() {
    int i;
    int arr[MAX], n;
    
    cout << "Enter number of elements: ";
    cin >> n;
    
    for (i = 1; i <= n; i++) {
        cout << "Enter element " << i << ": ";
        cin >> arr[i];
    }
    
    cout << "Entered list is: " << endl;
    display(arr, n);

    heap_sort(arr, n);

    cout << "Sorted list is: " << endl;
    display(arr, n);
    return 0;
}

void heap_sort(int arr[], int size) {
    int max;
    buildHeap(arr, size);
    cout << "Heap is: ";
    display(arr, size);
        
    while (size > 1) {
        max = del_root(arr, &size);
        arr[size + 1] = max;
    }
}

void buildHeap(int arr[], int size) {
    int i;
    for (i = size / 2; i >= 1; i--) {
        restoreDown(arr, i, size);
    }
}

int del_root(int arr[], int *size) {
    int max = arr[1];
    arr[1] = arr[*size];
    (*size)--;
    restoreDown(arr, 1, *size);
    return max;
} 

void restoreDown(int arr[], int i, int size) {
    int left = 2 * i, right = left + 1;
    int num = arr[i];
    
    while (right <= size) {
        if (num >= arr[left] && num >= arr[right]) {
            arr[i] = num;
            return;
        } else if (arr[left] > arr[right]) {
            arr[i] = arr[left]; 
            i = left;
        } else {
            arr[i] = arr[right];
            i = right;
        }
        left = 2 * i;
        right = left + 1;
    }

    if (left == size && num < arr[left]) { 
        arr[i] = arr[left];
        i = left;
    }
    arr[i] = num;
} 

void display(int arr[], int n) {       
    for (int i = 1; i <= n; i++) {
        cout << arr[i] << "  ";
    }
    cout << endl;
}
